src/bert_layers/attention.py

```diff
@@ -333,6 +333,112 @@ def forward(
             attn = bert_padding.unpad_input_only(attn, torch.squeeze(attn_mask) == 1)
 
         return self.out_drop(self.Wo(attn.view(bs, dim)))
+    
+
+class FlexBertUnpadParallelAttention(FlexBertAttentionBase):
+    """Computes the output of the multi-headed self parallel attention on a batch of unpadded sequences
+
+    If Flash Attention 2 is installed, this module uses Flash Attention to improve throughput.
+    If Flash Attention 2 is not installed, the implementation will use PyTorch's SDPA kernel,
+    which requires padding and unpadding inputs, adding some overhead.
+
+    See `forward` method for additional detail.
+    """
+
+    def __init__(self, config: FlexBertConfig):
+        super().__init__()
+        if config.hidden_size % config.num_attention_heads != 0 and not hasattr(config, "embedding_size"):
+            raise ValueError(
+                f"The hidden size ({config.hidden_size}) is not a multiple of the number of attention "
+                f"heads ({config.num_attention_heads})"
+            )
+
+        self.num_attention_heads = config.num_attention_heads
+        self.attn_head_size = int(config.hidden_size / config.num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attn_head_size
+        self.p_dropout = config.attention_probs_dropout_prob
+        # Compute QKV and FF outputs at once
+        self.Wqkvff = nn.Linear(config.hidden_size, 3 * self.all_head_size + config.intermediate_size, bias=config.attn_qkv_bias)
```

Comment:
FWIW I don't think we will ever use bias, so I would remove it entirely for the sake of simplicity

Reply:
All the linear layers currently have a bias option. I'd leave it in for consistency.